docs/appengine/firebase/tictactoe/main.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2016 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Sample tictactoe implements a tic-tac-toe game using Firebase as a communication mechanism.
+package tictactoe
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/zabawaba99/firego"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/datastore"
+	"google.golang.org/appengine/log"
+	"google.golang.org/appengine/user"
+)
+
+func init() {
+	http.HandleFunc("/", indexHandler)
+	http.HandleFunc("/opened", openedHandler)
+	http.HandleFunc("/move", moveHandler)
+	http.HandleFunc("/delete", deleteHandler)
+}
+
+var tmpl = template.Must(template.ParseFiles("template.html"))
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+
+	config, err := ioutil.ReadFile("firebase_config.html")
+	if err != nil {
+		handleError(w, r, "Could not read config file", err)
+		return
+	}
+
+	u := user.Current(ctx)
+
+	var game *Game
+	if gameID := r.FormValue("g"); gameID == "" {
+		game = NewGame()
+		game.K = datastore.NewKey(ctx, "Game", u.ID, 0, nil)
+		game.UserX = u.ID
+
+		if _, err := datastore.Put(ctx, game.K, game); err != nil {
+			handleError(w, r, "Could not start game", err)
+			return
+		}
+	} else {
+		// Existing game, join it.
+		if err := datastore.RunInTransaction(ctx, func(ctx context.Context) error {
+			g, err := gameFromRequest(r)
+			if err != nil {
+				return err
+			}
+			game = g
+			if game.UserO == "" {
```

Comment:
It might also mean User X just re-joined the game. Could check for that, but probably not worth it.